<a href="https://colab.research.google.com/github/0xjessie21/data-science-2026/blob/main/Pertemuan2_MOHAMMAD_RIYAN_SYAIFUNAHAR_240401010292.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pertemuan 2 — Struktur Data Python, NumPy & Pandas
**Nama:** MOHAMMAD RIYAN SYAIFUNAHAR  
**NIM:** 240401010292  
**Mata Kuliah:** Pengantar Data Science

## Persiapan — Import Library

In [1]:
import pandas as pd
import numpy as np

## 4.3 Pandas DataFrame — Latihan Dasar

In [3]:
# Membuat DataFrame dari dictionary
data = {
    'nama': ['Ali', 'Budi', 'Caca', 'Dina'],
    'usia': [22, 25, 21, 23],
    'kota': ['Jakarta', 'Bandung', 'Surabaya', 'Medan'],
    'ipk' : [3.8, 3.5, 3.9, 3.7]
}
df_latihan = pd.DataFrame(data)

# Cek informasi dasar DataFrame
print(df_latihan.shape)    # jumlah baris & kolom
print(df_latihan.dtypes)   # tipe data tiap kolom
print(df_latihan.columns)  # nama kolom
df_latihan

(4, 4)
nama     object
usia      int64
kota     object
ipk     float64
dtype: object
Index(['nama', 'usia', 'kota', 'ipk'], dtype='object')


,nama,usia,kota,ipk
0,Ali,22,Jakarta,3.8
1,Budi,25,Bandung,3.5
2,Caca,21,Surabaya,3.9
3,Dina,23,Medan,3.7


## 4.4 Memuat Dataset Titanic dari URL

In [4]:
# Load dataset dari URL GitHub
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)

# Eksplorasi standar setelah load
df.head()     # 5 baris pertama

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [5]:
df.info()      # tipe data & jumlah non-null per kolom

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [6]:
df.describe()  # statistik: mean, std, min, max, dll

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## 4.5 Filtering dan Seleksi Data

In [7]:
# Filter penumpang yang selamat
selamat = df[df['Survived'] == 1]
print("Jumlah selamat:", len(selamat))

# Filter penumpang wanita
wanita = df[df['Sex'] == 'female']
print("Jumlah wanita:", len(wanita))

Jumlah selamat: 342
Jumlah wanita: 314


In [8]:
# Filter wanita DAN kelas 1 (& = AND)
wanita_k1 = df[(df['Sex'] == 'female') & (df['Pclass'] == 1)]
print("Jumlah wanita kelas 1:", len(wanita_k1))
print(f"Rata-rata usia: {wanita_k1['Age'].mean():.2f} tahun")

# Lihat kolom tertentu saja
wanita_k1[['Name', 'Age', 'Survived']].head(10)

Jumlah wanita kelas 1: 94
Rata-rata usia: 34.61 tahun


,Name,Age,Survived
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,1
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,1
11,"Bonnell, Miss. Elizabeth",58.0,1
31,"Spencer, Mrs. William Augustus (Marie Eugenie)",NaN,1
52,"Harper, Mrs. Henry Sleeper (Myna Haxtun)",49.0,1
61,"Icard, Miss. Amelie",38.0,1
88,"Fortune, Miss. Mabel Helen",23.0,1
136,"Newsom, Miss. Helen Monypeny",19.0,1
151,"Pears, Mrs. Thomas (Edith Wearne)",22.0,1
166,"Chibnall, Mrs. (Edith Martha Bowerman)",NaN,1


In [9]:
# .loc[baris, kolom] — berbasis label
df.loc[df['Pclass'] == 1, ['Name', 'Age', 'Sex', 'Survived']].head()

,Name,Age,Sex,Survived
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,female,1
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,female,1
6,"McCarthy, Mr. Timothy J",54.0,male,0
11,"Bonnell, Miss. Elizabeth",58.0,female,1
23,"Sloper, Mr. William Thompson",28.0,male,1


## 4.6 Groupby dan Agregasi

In [10]:
# Rata-rata Survived per kelas (pola dasar groupby)
df.groupby('Pclass')['Survived'].mean().round(3)

,Survived
Pclass,
1,0.630
2,0.473
3,0.242


In [11]:
# Agregasi multiple: total, selamat, % selamat, rata-rata usia
hasil = df.groupby('Pclass').agg(
    total       = ('PassengerId', 'count'),
    selamat     = ('Survived',    'sum'),
    pct_selamat = ('Survived',    'mean'),
    rata_usia   = ('Age',         'mean')
)
hasil['pct_selamat'] = (hasil['pct_selamat'] * 100).round(2)
hasil['rata_usia']   = hasil['rata_usia'].round(1)

print(hasil)

# Kelas mana yang paling tinggi survival rate-nya?
terbaik = hasil['pct_selamat'].idxmax()
print(f"\nKelas dengan survival tertinggi: Kelas {terbaik}")

        total  selamat  pct_selamat  rata_usia
Pclass                                        
1         216      136        62.96       38.2
2         184       87        47.28       29.9
3         491      119        24.24       25.1

Kelas dengan survival tertinggi: Kelas 1


In [12]:
# Survival rate per kelas DAN jenis kelamin
df.groupby(['Pclass', 'Sex'])['Survived']\
  .mean().unstack().round(3)

Sex,female,male
Pclass,,
1,0.968,0.369
2,0.921,0.157
3,0.500,0.135


---
# Pertanyaan Latihan (Q1 – Q6)
**Dataset:** Titanic (891 penumpang, sumber: GitHub datasciencedojo)

## Q1 — Jumlah Baris, Kolom, dan Nama Kolom

In [ ]:
# Jumlah baris dan kolom
print("Shape (baris, kolom):", df.shape)

# Nama-nama kolom
print("\nDaftar kolom:")
print(df.columns.tolist())

# Info lengkap tipe data per kolom
df.info()

**Temuan Q1:**  
Dataset Titanic memiliki **891 baris** (penumpang) dan **12 kolom**.  
Kolom yang tersedia: `PassengerId`, `Survived`, `Pclass`, `Name`, `Sex`, `Age`, `SibSp`, `Parch`, `Ticket`, `Fare`, `Cabin`, `Embarked`.  
Kolom mencakup identitas penumpang, status keselamatan, kelas tiket, demografi, informasi keluarga, dan tiket.

## Q2 — Missing Values per Kolom

In [ ]:
# Jumlah missing values per kolom
missing = df.isnull().sum()

# Persentase missing values
persen = (df.isnull().sum() / len(df) * 100).round(2)

# Gabung menjadi tabel
missing_df = pd.DataFrame({
    'Jumlah Missing': missing,
    'Persentase (%)': persen
})

# Tampilkan hanya kolom yang punya missing values
print("Kolom dengan missing values:")
missing_df[missing_df['Jumlah Missing'] > 0]

**Temuan Q2:**  
Terdapat **3 kolom** dengan missing values:  
- `Age`: 177 data hilang (19.87%) — cukup signifikan, perlu strategi imputasi  
- `Cabin`: 687 data hilang (77.10%) — sangat banyak, sering diabaikan dalam analisis  
- `Embarked`: 2 data hilang (0.22%) — sangat sedikit, mudah diisi dengan modus

## Q3 — Persentase Penumpang yang Selamat

In [ ]:
# Menggunakan NumPy
survived_pct = np.mean(df['Survived']) * 100
print(f"Persentase selamat      : {survived_pct:.2f}%")
print(f"Persentase tidak selamat: {100 - survived_pct:.2f}%")
print(f"Jumlah selamat          : {df['Survived'].sum()} orang")
print(f"Jumlah tidak selamat    : {(df['Survived'] == 0).sum()} orang")

# Menggunakan Pandas (alternatif)
print("\nDistribusi Survived (Pandas):")
print((df['Survived'].value_counts(normalize=True) * 100).round(2))

**Temuan Q3:**  
Hanya **38.38%** (342 dari 891 penumpang) yang berhasil selamat.  
Lebih dari separuh penumpang yaitu **61.62%** (549 orang) tidak selamat dari tragedi Titanic.

## Q4 — Filter Penumpang Wanita Kelas 1

In [ ]:
# Filter boolean indexing: Sex == female DAN Pclass == 1
wanita_k1 = df[(df['Sex'] == 'female') & (df['Pclass'] == 1)]

print(f"Jumlah penumpang wanita kelas 1 : {len(wanita_k1)} orang")
print(f"Rata-rata usia                  : {wanita_k1['Age'].mean():.2f} tahun")
print(f"Jumlah yang selamat             : {wanita_k1['Survived'].sum()} orang")
print(f"Survival rate                   : {wanita_k1['Survived'].mean()*100:.2f}%")

# Tampilkan 10 baris pertama
wanita_k1[['Name', 'Age', 'Survived']].head(10)

**Temuan Q4:**  
Terdapat **94 penumpang wanita di kelas 1** dengan rata-rata usia **34.61 tahun**.  
Kelompok ini memiliki survival rate sangat tinggi karena mendapat prioritas evakuasi sesuai kebijakan *"women and children first"*.

## Q5 — Tingkat Keselamatan per Kelas Penumpang

In [ ]:
# Groupby Pclass dengan agregasi lengkap
hasil = df.groupby('Pclass').agg(
    total       = ('PassengerId', 'count'),
    selamat     = ('Survived',    'sum'),
    pct_selamat = ('Survived',    'mean'),
    rata_usia   = ('Age',         'mean')
)
hasil['pct_selamat'] = (hasil['pct_selamat'] * 100).round(2)
hasil['rata_usia']   = hasil['rata_usia'].round(1)

print("Tingkat Keselamatan per Kelas Penumpang:")
print(hasil)

# Kelas dengan survival rate tertinggi
terbaik = hasil['pct_selamat'].idxmax()
print(f"\nKelas dengan survival tertinggi: Kelas {terbaik} ({hasil.loc[terbaik, 'pct_selamat']}%)")

**Temuan Q5:**  
**Kelas 1 (Pertama)** memiliki tingkat keselamatan tertinggi yaitu **62.96%**.  
Semakin rendah kelas penumpang, semakin kecil peluang selamat:  
- Kelas 1: 62.96% selamat  
- Kelas 2: 47.28% selamat  
- Kelas 3: 24.24% selamat  

Hal ini mencerminkan ketidaksetaraan akses ke sekoci penyelamat berdasarkan kelas sosial ekonomi.

## Q6 — Analisis Tambahan: Survival Rate Berdasarkan Jenis Kelamin dan Kelas

In [ ]:
# Survival rate per jenis kelamin dan kelas (groupby multi-kolom)
print("=== Survival Rate per Kelas & Jenis Kelamin (%) ===")
tabel = df.groupby(['Pclass', 'Sex'])['Survived']\
          .mean().unstack().round(3) * 100
print(tabel)

print("\n=== Survival Rate per Jenis Kelamin ===")
gender = df.groupby('Sex')['Survived'].agg(['sum', 'mean', 'count'])
gender.rename(columns={'sum': 'Selamat', 'mean': 'Rate', 'count': 'Total'}, inplace=True)
gender['Rate (%)'] = (gender['Rate'] * 100).round(2)
print(gender[['Total', 'Selamat', 'Rate (%)']])

**Temuan Q6:**  
**Jenis kelamin adalah faktor terkuat** yang mempengaruhi keselamatan penumpang Titanic:  
- Wanita kelas 1: **96.8%** selamat  
- Wanita kelas 2: **92.1%** selamat  
- Wanita kelas 3: **50.0%** selamat  
- Pria kelas 1: **36.9%** selamat  
- Pria kelas 2: **15.7%** selamat  
- Pria kelas 3: **13.5%** selamat  

Kombinasi **gender + kelas sosial** sangat menentukan peluang hidup.  
Ini mencerminkan dua faktor utama: (1) prioritas evakuasi *"women and children first"*, dan (2) ketimpangan akses ke dek sekoci berdasarkan kelas kabin.